base/channels.jl

```diff
@@ -454,6 +454,118 @@ function put_unbuffered(c::Channel, v)
     return v
 end
 
+"""
+    append!(c::Channel, iter)
+
+Append all items in `iter` to the channel `c`. If the channel is buffered, this operation requires
+fewer `lock` operations than individual `put!`s. Blocks if the channel is full.
+
+# Examples
+
+```jldoctest
+julia> c = Channel(3);
+
+julia> append!(c, 1:3);
+
+julia> take!(c)
+1
+
+julia> take!(c)
+2
+
+julia> take!(c)
+3
+```
+
+!!! compat "Julia 1.12"
+    Requires at least Julia 1.12.
+"""
+function append!(c::Channel, iter::T) where {T}
+    if hasmethod(length, Tuple{T})
+        len = length(iter)
+        # shortcircuit for small short objects
+        if len == 0
+            return c
+        elseif len == 1
+            put!(c, @inbounds vec[begin])
+            return c
+        end
+    end
+
+    return isbuffered(c) ? append_buffered(c, iter) : append_unbuffered(c, iter)
+end
+function append!(c1::Channel, c2::Channel{T}) where {T}
+    buff_len = max(c1.sz_max, c2.sz_max)
+    if buff_len == 0
+        return append_unbuffered(c1, c2)
+    end
+
+    buff = Vector{T}(undef, buff_len)
+    while isopen(c2) || isready(c2)
+        take!(c2, buff_len, buff)
+        append_buffered(c1, buff)
+    end
+    return c1
+end
+
+function append_unbuffered(c::Channel, iter)
+    for v in iter
+        put!(c, v)
+    end
+    return c
+end
+
+function append_buffered(c::Channel{T}, iter::I) where {T,I}
```

Comment:
It's more idiomatic in Julia to rely on the `Base.IteratorSize` trait rather than checking `haslength`

Reply:
I am not clear on how to deal with arbitrary types while relying on `Base.IteratorSize`. For example

```julia
julia> using DataFrames

julia> df = DataFrame()
0×0 DataFrame

julia> Base.IteratorSize(df)
Base.HasLength()

julia> length(df)
ERROR: MethodError: no method matching length(::DataFrame)
The function `length` exists, but no method is defined for this combination of argument types.       
You may need to implement the `length` method or define `IteratorSize` for this type to be `SizeUnknown`.

Closest candidates are:
  length(::Base.AsyncGenerator)
   @ Base asyncmap.jl:392
  length(::CompositeException)
   @ Base task.jl:51
  length(::Markdown.MD)
   @ Markdown C:\Users\mrufsvold\.julia\juliaup\julia-nightly\share\julia\stdlib\v1.12\Markdown\src\parse\parse.jl:37
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[32]:1
```
So getting `Base.HasLength()` back does not imply that you can safely query `length(x)`. Inversely:

```julia
julia> m = [1 2; 3 4]
2×2 Matrix{Int64}:
 1  2
 3  4

julia> Base.IteratorSize(m)
Base.HasShape{2}()

julia> length(m)
4
```

Reply:
It's only supposed to work with iterable types, right? And any type that defines `iterate` and is `HasLength` should also define `length`. In your example, it errors with a data frame, but that isn't supposed to work, because data frames cannot be iterated.

Reply:
done